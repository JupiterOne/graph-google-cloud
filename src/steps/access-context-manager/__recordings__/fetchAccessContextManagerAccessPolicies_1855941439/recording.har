{
  "log": {
    "_recordingName": "fetchAccessContextManagerAccessPolicies",
    "creator": {
      "comment": "persister:JupiterOneIntegationFSPersister",
      "name": "Polly.JS",
      "version": "4.3.0"
    },
    "entries": [
      {
        "_id": "acea721c8193b51ced888cae721cc423",
        "_order": 0,
        "cache": {},
        "request": {
          "bodySize": 739,
          "cookies": [],
          "headers": [
            {
              "_fromType": "array",
              "name": "content-type",
              "value": "application/x-www-form-urlencoded"
            },
            {
              "_fromType": "array",
              "name": "accept",
              "value": "application/json"
            },
            {
              "_fromType": "array",
              "name": "content-length",
              "value": "739"
            },
            {
              "_fromType": "array",
              "name": "user-agent",
              "value": "node-fetch/1.0 (+https://github.com/bitinn/node-fetch)"
            },
            {
              "_fromType": "array",
              "name": "accept-encoding",
              "value": "gzip,deflate"
            },
            {
              "_fromType": "array",
              "name": "connection",
              "value": "close"
            },
            {
              "name": "host",
              "value": "www.googleapis.com"
            }
          ],
          "headersSize": 300,
          "httpVersion": "HTTP/1.1",
          "method": "POST",
          "postData": {
            "mimeType": "application/x-www-form-urlencoded",
            "params": [],
            "text": "[REDACTED]"
          },
          "queryString": [],
          "url": "https://www.googleapis.com/oauth2/v4/token"
        },
        "response": {
          "bodySize": 572,
          "content": {
            "mimeType": "application/json; charset=UTF-8",
            "size": 572,
            "text": "{\"access_token\":\"[REDACTED]\",\"expires_in\":9999,\"token_type\":\"Bearer\"}"
          },
          "cookies": [],
          "headers": [
            {
              "name": "content-type",
              "value": "application/json; charset=UTF-8"
            },
            {
              "name": "vary",
              "value": "Origin, X-Origin, Referer"
            },
            {
              "name": "date",
              "value": "Wed, 02 Jun 2021 14:22:23 GMT"
            },
            {
              "name": "server",
              "value": "scaffolding on HTTPServer2"
            },
            {
              "name": "cache-control",
              "value": "private"
            },
            {
              "name": "x-xss-protection",
              "value": "0"
            },
            {
              "name": "x-frame-options",
              "value": "SAMEORIGIN"
            },
            {
              "name": "x-content-type-options",
              "value": "nosniff"
            },
            {
              "name": "alt-svc",
              "value": "h3-29=\":443\"; ma=2592000,h3-T051=\":443\"; ma=2592000,h3-Q050=\":443\"; ma=2592000,h3-Q046=\":443\"; ma=2592000,h3-Q043=\":443\"; ma=2592000,quic=\":443\"; ma=2592000; v=\"46,43\""
            },
            {
              "name": "connection",
              "value": "close"
            }
          ],
          "headersSize": 511,
          "httpVersion": "HTTP/1.1",
          "redirectURL": "",
          "status": 200,
          "statusText": "OK"
        },
        "startedDateTime": "2021-06-02T14:22:23.123Z",
        "time": 212,
        "timings": {
          "blocked": -1,
          "connect": -1,
          "dns": -1,
          "receive": 0,
          "send": 0,
          "ssl": -1,
          "wait": 212
        }
      },
      {
        "_id": "e7ae7a9d3285d95d4330a7c239aa61b7",
        "_order": 0,
        "cache": {},
        "request": {
          "bodySize": 0,
          "cookies": [],
          "headers": [
            {
              "_fromType": "array",
              "name": "x-goog-api-client",
              "value": "gdcl/5.0.2 gl-node/14.15.4 auth/7.1.0"
            },
            {
              "_fromType": "array",
              "name": "accept-encoding",
              "value": "gzip"
            },
            {
              "_fromType": "array",
              "name": "user-agent",
              "value": "google-api-nodejs-client/5.0.2 (gzip)"
            },
            {
              "_fromType": "array",
              "name": "authorization",
              "value": "[REDACTED]"
            },
            {
              "_fromType": "array",
              "name": "accept",
              "value": "application/json"
            },
            {
              "_fromType": "array",
              "name": "connection",
              "value": "close"
            },
            {
              "name": "host",
              "value": "accesscontextmanager.googleapis.com"
            }
          ],
          "headersSize": 581,
          "httpVersion": "HTTP/1.1",
          "method": "GET",
          "queryString": [
            {
              "name": "parent",
              "value": "organizations/958457776463"
            }
          ],
          "url": "https://accesscontextmanager.googleapis.com/v1/accessPolicies?parent=organizations%2F958457776463"
        },
        "response": {
          "bodySize": 369,
          "content": {
            "mimeType": "application/json; charset=UTF-8",
            "size": 369,
            "text": "{\"accessPolicies\":[{\"name\":\"accessPolicies/368605946207\",\"parent\":\"organizations/958457776463\",\"title\":\"default policy\",\"etag\":\"64186ca0c82b01b4\"}]}"
          },
          "cookies": [],
          "headers": [
            {
              "name": "content-type",
              "value": "application/json; charset=UTF-8"
            },
            {
              "name": "vary",
              "value": "Origin, X-Origin, Referer"
            },
            {
              "name": "date",
              "value": "Wed, 02 Jun 2021 14:22:24 GMT"
            },
            {
              "name": "server",
              "value": "ESF"
            },
            {
              "name": "cache-control",
              "value": "private"
            },
            {
              "name": "x-xss-protection",
              "value": "0"
            },
            {
              "name": "x-frame-options",
              "value": "SAMEORIGIN"
            },
            {
              "name": "x-content-type-options",
              "value": "nosniff"
            },
            {
              "name": "alt-svc",
              "value": "h3-29=\":443\"; ma=2592000,h3-T051=\":443\"; ma=2592000,h3-Q050=\":443\"; ma=2592000,h3-Q046=\":443\"; ma=2592000,h3-Q043=\":443\"; ma=2592000,quic=\":443\"; ma=2592000; v=\"46,43\""
            },
            {
              "name": "connection",
              "value": "close"
            }
          ],
          "headersSize": 488,
          "httpVersion": "HTTP/1.1",
          "redirectURL": "",
          "status": 200,
          "statusText": "OK"
        },
        "startedDateTime": "2021-06-02T14:22:23.374Z",
        "time": 829,
        "timings": {
          "blocked": -1,
          "connect": -1,
          "dns": -1,
          "receive": 0,
          "send": 0,
          "ssl": -1,
          "wait": 829
        }
      }
    ],
    "pages": [],
    "version": "1.2"
  }
}
