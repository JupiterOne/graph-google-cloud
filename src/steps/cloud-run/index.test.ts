import { createMockStepExecutionContext } from '@jupiterone/integration-sdk-testing';
import {
  fetchCloudRunConfigurations,
  fetchCloudRunRoutes,
  fetchCloudRunServices,
} from '.';
import { integrationConfig } from '../../../test/config';
import { Recording, setupGoogleCloudRecording } from '../../../test/recording';
import { IntegrationConfig } from '../../types';
import {
  ENTITY_TYPE_CLOUD_RUN_CONFIGURATION,
  ENTITY_TYPE_CLOUD_RUN_ROUTE,
  ENTITY_TYPE_CLOUD_RUN_SERVICE,
  RELATIONSHIP_TYPE_CLOUD_RUN_SERVICE_MANAGES_CONFIGURATION,
  RELATIONSHIP_TYPE_CLOUD_RUN_SERVICE_MANAGES_ROUTE,
} from './constants';

describe('#fetchCloudRunServices', () => {
  let recording: Recording;

  beforeEach(() => {
    recording = setupGoogleCloudRecording({
      directory: __dirname,
      name: 'fetchCloudRunServices',
    });
  });

  afterEach(async () => {
    await recording.stop();
  });

  test('should collect data', async () => {
    const context = createMockStepExecutionContext<IntegrationConfig>({
      instanceConfig: integrationConfig,
    });

    await fetchCloudRunServices(context);

    expect({
      numCollectedEntities: context.jobState.collectedEntities.length,
      numCollectedRelationships: context.jobState.collectedRelationships.length,
      collectedEntities: context.jobState.collectedEntities,
      collectedRelationships: context.jobState.collectedRelationships,
      encounteredTypes: context.jobState.encounteredTypes,
    }).toMatchSnapshot();

    expect(
      context.jobState.collectedEntities.filter(
        (e) => e._type === ENTITY_TYPE_CLOUD_RUN_SERVICE,
      ),
    ).toMatchGraphObjectSchema({
      _class: ['Service'],
      schema: {
        additionalProperties: true,
        properties: {
          _type: { const: 'google_cloud_run_service' },
          _rawData: {
            type: 'array',
            items: { type: 'object' },
          },

          name: { type: 'string' },
          displayName: { type: 'string' },
          category: {
            type: 'array',
            items: { type: 'string' },
          },
          createdOn: { type: 'number' },
          webLink: { type: 'string' },
        },
      },
    });
  });
});

describe('#fetchCloudRunRoutes', () => {
  let recording: Recording;

  beforeEach(() => {
    recording = setupGoogleCloudRecording({
      directory: __dirname,
      name: 'fetchCloudRunRoutes',
    });
  });

  afterEach(async () => {
    await recording.stop();
  });

  test('should collect data', async () => {
    const context = createMockStepExecutionContext<IntegrationConfig>({
      instanceConfig: integrationConfig,
    });

    await fetchCloudRunServices(context);
    await fetchCloudRunRoutes(context);

    expect({
      numCollectedEntities: context.jobState.collectedEntities.length,
      numCollectedRelationships: context.jobState.collectedRelationships.length,
      collectedEntities: context.jobState.collectedEntities,
      collectedRelationships: context.jobState.collectedRelationships,
      encounteredTypes: context.jobState.encounteredTypes,
    }).toMatchSnapshot();

    expect(
      context.jobState.collectedEntities.filter(
        (e) => e._type === ENTITY_TYPE_CLOUD_RUN_SERVICE,
      ),
    ).toMatchGraphObjectSchema({
      _class: ['Service'],
      schema: {
        additionalProperties: true,
        properties: {
          _type: { const: 'google_cloud_run_service' },
          _rawData: {
            type: 'array',
            items: { type: 'object' },
          },

          name: { type: 'string' },
          displayName: { type: 'string' },
          category: {
            type: 'array',
            items: { type: 'string' },
          },
          createdOn: { type: 'number' },
          webLink: { type: 'string' },
        },
      },
    });

    expect(
      context.jobState.collectedEntities.filter(
        (e) => e._type === ENTITY_TYPE_CLOUD_RUN_ROUTE,
      ),
    ).toMatchGraphObjectSchema({
      _class: ['Configuration'],
      schema: {
        additionalProperties: false,
        properties: {
          _type: { const: 'google_cloud_run_route' },
          _rawData: {
            type: 'array',
            items: { type: 'object' },
          },
          name: { type: 'string' },
          displayName: { type: 'string' },
          url: { type: 'string' },
          createdOn: { type: 'number' },
        },
      },
    });

    expect(
      context.jobState.collectedRelationships.filter(
        (e) => e._type === RELATIONSHIP_TYPE_CLOUD_RUN_SERVICE_MANAGES_ROUTE,
      ),
    ).toMatchDirectRelationshipSchema({
      schema: {
        properties: {
          _class: { const: 'MANAGES' },
          _type: { const: 'google_cloud_run_service_manages_route' },
        },
      },
    });
  });
});

describe('#fetchCloudRunConfigurations', () => {
  let recording: Recording;

  beforeEach(() => {
    recording = setupGoogleCloudRecording({
      directory: __dirname,
      name: 'fetchCloudRunConfigurations',
    });
  });

  afterEach(async () => {
    await recording.stop();
  });

  test('should collect data', async () => {
    const context = createMockStepExecutionContext<IntegrationConfig>({
      instanceConfig: integrationConfig,
    });

    await fetchCloudRunServices(context);
    await fetchCloudRunConfigurations(context);

    expect({
      numCollectedEntities: context.jobState.collectedEntities.length,
      numCollectedRelationships: context.jobState.collectedRelationships.length,
      collectedEntities: context.jobState.collectedEntities,
      collectedRelationships: context.jobState.collectedRelationships,
      encounteredTypes: context.jobState.encounteredTypes,
    }).toMatchSnapshot();

    expect(
      context.jobState.collectedEntities.filter(
        (e) => e._type === ENTITY_TYPE_CLOUD_RUN_SERVICE,
      ),
    ).toMatchGraphObjectSchema({
      _class: ['Service'],
      schema: {
        additionalProperties: true,
        properties: {
          _type: { const: 'google_cloud_run_service' },
          _rawData: {
            type: 'array',
            items: { type: 'object' },
          },

          name: { type: 'string' },
          displayName: { type: 'string' },
          category: {
            type: 'array',
            items: { type: 'string' },
          },
          createdOn: { type: 'number' },
          webLink: { type: 'string' },
        },
      },
    });

    expect(
      context.jobState.collectedEntities.filter(
        (e) => e._type === ENTITY_TYPE_CLOUD_RUN_CONFIGURATION,
      ),
    ).toMatchGraphObjectSchema({
      _class: ['Configuration'],
      schema: {
        additionalProperties: true,
        properties: {
          _type: { const: 'google_cloud_run_configuration' },
          _rawData: {
            type: 'array',
            items: { type: 'object' },
          },
          name: { type: 'string' },
          displayName: { type: 'string' },
          apiVersion: { type: 'string' },
          createdOn: { type: 'number' },
        },
      },
    });

    expect(
      context.jobState.collectedRelationships.filter(
        (e) =>
          e._type === RELATIONSHIP_TYPE_CLOUD_RUN_SERVICE_MANAGES_CONFIGURATION,
      ),
    ).toMatchDirectRelationshipSchema({
      schema: {
        properties: {
          _class: { const: 'MANAGES' },
          _type: { const: 'google_cloud_run_service_manages_configuration' },
        },
      },
    });
  });
});
