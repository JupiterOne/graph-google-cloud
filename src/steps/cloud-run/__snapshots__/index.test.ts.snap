// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`#fetchCloudRunConfigurations should collect data 1`] = `
{
  "collectedEntities": [
    {
      "_class": [
        "Service",
      ],
      "_key": "projects/j1-gc-integration-dev-v2/locations/us-central1/services/example-cloud-run-service",
      "_rawData": [
        {
          "name": "default",
          "rawData": {
            "apiVersion": "serving.knative.dev/v1",
            "kind": "Service",
            "metadata": {
              "annotations": {
                "client.knative.dev/user-image": "us-docker.pkg.dev/cloudrun/container/hello",
                "run.googleapis.com/client-name": "cloud-console",
                "run.googleapis.com/ingress": "all",
                "run.googleapis.com/ingress-status": "all",
                "run.googleapis.com/launch-stage": "BETA",
                "serving.knative.dev/creator": "user@email.com",
                "serving.knative.dev/lastModifier": "user@email.com",
              },
              "creationTimestamp": "2021-03-09T20:41:29.622532Z",
              "generation": 1,
              "labels": {
                "cloud.googleapis.com/location": "us-central1",
              },
              "name": "example-cloud-run-service",
              "namespace": "711888229551",
              "resourceVersion": "AAW9IJgvqVc",
              "selfLink": "/apis/serving.knative.dev/v1/namespaces/711888229551/services/example-cloud-run-service",
              "uid": "fce2d165-9ccb-4b6c-9204-67362d41aadd",
            },
            "spec": {
              "template": {
                "metadata": {
                  "annotations": {
                    "autoscaling.knative.dev/maxScale": "100",
                    "run.googleapis.com/client-name": "cloud-console",
                    "run.googleapis.com/sandbox": "gvisor",
                  },
                  "name": "example-cloud-run-service-00001-mub",
                },
                "spec": {
                  "containerConcurrency": 80,
                  "containers": [
                    {
                      "image": "us-docker.pkg.dev/cloudrun/container/hello",
                      "ports": [
                        {
                          "containerPort": 8080,
                        },
                      ],
                      "resources": {
                        "limits": {
                          "cpu": "1000m",
                          "memory": "256Mi",
                        },
                      },
                    },
                  ],
                  "serviceAccountName": "711888229551-compute@developer.gserviceaccount.com",
                  "timeoutSeconds": 300,
                },
              },
              "traffic": [
                {
                  "latestRevision": true,
                  "percent": 100,
                },
              ],
            },
            "status": {
              "address": {
                "url": "https://example-cloud-run-service-rhbpqun2qa-uc.a.run.app",
              },
              "conditions": [
                {
                  "lastTransitionTime": "2021-03-09T20:42:53.416791Z",
                  "status": "True",
                  "type": "Ready",
                },
                {
                  "lastTransitionTime": "2021-03-09T20:42:53.211458Z",
                  "status": "True",
                  "type": "ConfigurationsReady",
                },
                {
                  "lastTransitionTime": "2021-03-09T20:42:53.416791Z",
                  "status": "True",
                  "type": "RoutesReady",
                },
              ],
              "latestCreatedRevisionName": "example-cloud-run-service-00001-mub",
              "latestReadyRevisionName": "example-cloud-run-service-00001-mub",
              "observedGeneration": 1,
              "traffic": [
                {
                  "latestRevision": true,
                  "percent": 100,
                  "revisionName": "example-cloud-run-service-00001-mub",
                },
              ],
              "url": "https://example-cloud-run-service-rhbpqun2qa-uc.a.run.app",
            },
          },
        },
      ],
      "_type": "google_cloud_run_service",
      "active": undefined,
      "category": [
        "infrastructure",
      ],
      "createdOn": 1615322489622,
      "displayName": "example-cloud-run-service",
      "function": [
        "workload-management",
      ],
      "name": "example-cloud-run-service",
      "webLink": "https://console.cloud.google.com/run/detail/us-central1/example-cloud-run-service",
    },
    {
      "_class": [
        "Configuration",
      ],
      "_key": "cloudrun_configuration:f190d809-4de1-4eb3-80bb-ad0ac9954f85",
      "_rawData": [
        {
          "name": "default",
          "rawData": {
            "apiVersion": "serving.knative.dev/v1",
            "kind": "Configuration",
            "metadata": {
              "annotations": {
                "client.knative.dev/user-image": "us-docker.pkg.dev/cloudrun/container/hello",
                "run.googleapis.com/client-name": "cloud-console",
                "run.googleapis.com/launch-stage": "BETA",
                "serving.knative.dev/creator": "user@email.com",
                "serving.knative.dev/lastModifier": "user@email.com",
              },
              "creationTimestamp": "2021-03-09T20:41:30.336883Z",
              "generation": 1,
              "labels": {
                "cloud.googleapis.com/location": "us-central1",
                "serving.knative.dev/route": "example-cloud-run-service",
                "serving.knative.dev/service": "example-cloud-run-service",
              },
              "name": "example-cloud-run-service",
              "namespace": "711888229551",
              "ownerReferences": [
                {
                  "apiVersion": "serving.knative.dev/v1",
                  "blockOwnerDeletion": true,
                  "controller": true,
                  "kind": "Service",
                  "name": "example-cloud-run-service",
                  "uid": "fce2d165-9ccb-4b6c-9204-67362d41aadd",
                },
              ],
              "resourceVersion": "AAW9IJgrZIw",
              "selfLink": "/apis/serving.knative.dev/v1/namespaces/711888229551/configurations/example-cloud-run-service",
              "uid": "f190d809-4de1-4eb3-80bb-ad0ac9954f85",
            },
            "spec": {
              "template": {
                "metadata": {
                  "annotations": {
                    "autoscaling.knative.dev/maxScale": "100",
                    "run.googleapis.com/client-name": "cloud-console",
                    "run.googleapis.com/sandbox": "gvisor",
                  },
                  "name": "example-cloud-run-service-00001-mub",
                },
                "spec": {
                  "containerConcurrency": 80,
                  "containers": [
                    {
                      "image": "us-docker.pkg.dev/cloudrun/container/hello",
                      "ports": [
                        {
                          "containerPort": 8080,
                        },
                      ],
                      "resources": {
                        "limits": {
                          "cpu": "1000m",
                          "memory": "256Mi",
                        },
                      },
                    },
                  ],
                  "serviceAccountName": "711888229551-compute@developer.gserviceaccount.com",
                  "timeoutSeconds": 300,
                },
              },
            },
            "status": {
              "conditions": [
                {
                  "lastTransitionTime": "2021-03-09T20:42:53.137036Z",
                  "status": "True",
                  "type": "Ready",
                },
              ],
              "latestCreatedRevisionName": "example-cloud-run-service-00001-mub",
              "latestReadyRevisionName": "example-cloud-run-service-00001-mub",
              "observedGeneration": 1,
            },
          },
        },
      ],
      "_type": "google_cloud_run_configuration",
      "active": undefined,
      "apiVersion": "serving.knative.dev/v1",
      "createdOn": 1615322490336,
      "displayName": "example-cloud-run-service",
      "metaClientName": "cloud-console",
      "metaCreator": "user@email.com",
      "metaLastModifier": "user@email.com",
      "metaLaunchStage": "BETA",
      "metaMaxScale": "100",
      "metaSandbox": "gvisor",
      "metaUserImage": "us-docker.pkg.dev/cloudrun/container/hello",
      "name": "example-cloud-run-service",
    },
  ],
  "collectedRelationships": [
    {
      "_class": "MANAGES",
      "_fromEntityKey": "projects/j1-gc-integration-dev-v2/locations/us-central1/services/example-cloud-run-service",
      "_key": "projects/j1-gc-integration-dev-v2/locations/us-central1/services/example-cloud-run-service|manages|cloudrun_configuration:f190d809-4de1-4eb3-80bb-ad0ac9954f85",
      "_toEntityKey": "cloudrun_configuration:f190d809-4de1-4eb3-80bb-ad0ac9954f85",
      "_type": "google_cloud_run_service_manages_configuration",
      "displayName": "MANAGES",
    },
  ],
  "encounteredTypes": [
    "google_cloud_run_service",
    "google_cloud_run_configuration",
    "google_cloud_run_service_manages_configuration",
  ],
  "numCollectedEntities": 2,
  "numCollectedRelationships": 1,
}
`;

exports[`#fetchCloudRunRoutes should collect data 1`] = `
{
  "collectedEntities": [
    {
      "_class": [
        "Service",
      ],
      "_key": "projects/j1-gc-integration-dev-v2/locations/us-central1/services/example-cloud-run-service",
      "_rawData": [
        {
          "name": "default",
          "rawData": {
            "apiVersion": "serving.knative.dev/v1",
            "kind": "Service",
            "metadata": {
              "annotations": {
                "client.knative.dev/user-image": "us-docker.pkg.dev/cloudrun/container/hello",
                "run.googleapis.com/client-name": "cloud-console",
                "run.googleapis.com/ingress": "all",
                "run.googleapis.com/ingress-status": "all",
                "run.googleapis.com/launch-stage": "BETA",
                "serving.knative.dev/creator": "user@email.com",
                "serving.knative.dev/lastModifier": "user@email.com",
              },
              "creationTimestamp": "2021-03-09T20:41:29.622532Z",
              "generation": 1,
              "labels": {
                "cloud.googleapis.com/location": "us-central1",
              },
              "name": "example-cloud-run-service",
              "namespace": "711888229551",
              "resourceVersion": "AAW9IJgvqVc",
              "selfLink": "/apis/serving.knative.dev/v1/namespaces/711888229551/services/example-cloud-run-service",
              "uid": "fce2d165-9ccb-4b6c-9204-67362d41aadd",
            },
            "spec": {
              "template": {
                "metadata": {
                  "annotations": {
                    "autoscaling.knative.dev/maxScale": "100",
                    "run.googleapis.com/client-name": "cloud-console",
                    "run.googleapis.com/sandbox": "gvisor",
                  },
                  "name": "example-cloud-run-service-00001-mub",
                },
                "spec": {
                  "containerConcurrency": 80,
                  "containers": [
                    {
                      "image": "us-docker.pkg.dev/cloudrun/container/hello",
                      "ports": [
                        {
                          "containerPort": 8080,
                        },
                      ],
                      "resources": {
                        "limits": {
                          "cpu": "1000m",
                          "memory": "256Mi",
                        },
                      },
                    },
                  ],
                  "serviceAccountName": "711888229551-compute@developer.gserviceaccount.com",
                  "timeoutSeconds": 300,
                },
              },
              "traffic": [
                {
                  "latestRevision": true,
                  "percent": 100,
                },
              ],
            },
            "status": {
              "address": {
                "url": "https://example-cloud-run-service-rhbpqun2qa-uc.a.run.app",
              },
              "conditions": [
                {
                  "lastTransitionTime": "2021-03-09T20:42:53.416791Z",
                  "status": "True",
                  "type": "Ready",
                },
                {
                  "lastTransitionTime": "2021-03-09T20:42:53.211458Z",
                  "status": "True",
                  "type": "ConfigurationsReady",
                },
                {
                  "lastTransitionTime": "2021-03-09T20:42:53.416791Z",
                  "status": "True",
                  "type": "RoutesReady",
                },
              ],
              "latestCreatedRevisionName": "example-cloud-run-service-00001-mub",
              "latestReadyRevisionName": "example-cloud-run-service-00001-mub",
              "observedGeneration": 1,
              "traffic": [
                {
                  "latestRevision": true,
                  "percent": 100,
                  "revisionName": "example-cloud-run-service-00001-mub",
                },
              ],
              "url": "https://example-cloud-run-service-rhbpqun2qa-uc.a.run.app",
            },
          },
        },
      ],
      "_type": "google_cloud_run_service",
      "active": undefined,
      "category": [
        "infrastructure",
      ],
      "createdOn": 1615322489622,
      "displayName": "example-cloud-run-service",
      "function": [
        "workload-management",
      ],
      "name": "example-cloud-run-service",
      "webLink": "https://console.cloud.google.com/run/detail/us-central1/example-cloud-run-service",
    },
    {
      "_class": [
        "Configuration",
      ],
      "_key": "cloudrun_route:f538487c-6e93-4038-9f6a-b7936cc4dd04",
      "_rawData": [
        {
          "name": "default",
          "rawData": {
            "apiVersion": "serving.knative.dev/v1",
            "kind": "Route",
            "metadata": {
              "annotations": {
                "client.knative.dev/user-image": "us-docker.pkg.dev/cloudrun/container/hello",
                "run.googleapis.com/client-name": "cloud-console",
                "run.googleapis.com/ingress": "all",
                "run.googleapis.com/ingress-status": "all",
                "serving.knative.dev/creator": "user@email.com",
                "serving.knative.dev/lastModifier": "user@email.com",
              },
              "creationTimestamp": "2021-03-09T20:41:30.309851Z",
              "generation": 1,
              "labels": {
                "cloud.googleapis.com/location": "us-central1",
                "serving.knative.dev/service": "example-cloud-run-service",
              },
              "name": "example-cloud-run-service",
              "namespace": "711888229551",
              "ownerReferences": [
                {
                  "apiVersion": "serving.knative.dev/v1",
                  "blockOwnerDeletion": true,
                  "controller": true,
                  "kind": "Service",
                  "name": "example-cloud-run-service",
                  "uid": "fce2d165-9ccb-4b6c-9204-67362d41aadd",
                },
              ],
              "resourceVersion": "AAW9IJgu/zw",
              "selfLink": "/apis/serving.knative.dev/v1/namespaces/711888229551/routes/example-cloud-run-service",
              "uid": "f538487c-6e93-4038-9f6a-b7936cc4dd04",
            },
            "spec": {
              "traffic": [
                {
                  "configurationName": "example-cloud-run-service",
                  "latestRevision": true,
                  "percent": 100,
                },
              ],
            },
            "status": {
              "address": {
                "url": "https://example-cloud-run-service-rhbpqun2qa-uc.a.run.app",
              },
              "conditions": [
                {
                  "lastTransitionTime": "2021-03-09T20:42:53.373244Z",
                  "status": "True",
                  "type": "Ready",
                },
              ],
              "observedGeneration": 1,
              "traffic": [
                {
                  "latestRevision": true,
                  "percent": 100,
                  "revisionName": "example-cloud-run-service-00001-mub",
                },
              ],
              "url": "https://example-cloud-run-service-rhbpqun2qa-uc.a.run.app",
            },
          },
        },
      ],
      "_type": "google_cloud_run_route",
      "active": undefined,
      "createdOn": 1615322490309,
      "displayName": "example-cloud-run-service",
      "name": "example-cloud-run-service",
      "url": "https://example-cloud-run-service-rhbpqun2qa-uc.a.run.app",
    },
  ],
  "collectedRelationships": [
    {
      "_class": "MANAGES",
      "_fromEntityKey": "projects/j1-gc-integration-dev-v2/locations/us-central1/services/example-cloud-run-service",
      "_key": "projects/j1-gc-integration-dev-v2/locations/us-central1/services/example-cloud-run-service|manages|cloudrun_route:f538487c-6e93-4038-9f6a-b7936cc4dd04",
      "_toEntityKey": "cloudrun_route:f538487c-6e93-4038-9f6a-b7936cc4dd04",
      "_type": "google_cloud_run_service_manages_route",
      "displayName": "MANAGES",
    },
  ],
  "encounteredTypes": [
    "google_cloud_run_service",
    "google_cloud_run_route",
    "google_cloud_run_service_manages_route",
  ],
  "numCollectedEntities": 2,
  "numCollectedRelationships": 1,
}
`;

exports[`#fetchCloudRunServices should collect data 1`] = `
{
  "collectedEntities": [
    {
      "_class": [
        "Service",
      ],
      "_key": "projects/j1-gc-integration-dev-v2/locations/us-central1/services/example-cloud-run-service",
      "_rawData": [
        {
          "name": "default",
          "rawData": {
            "apiVersion": "serving.knative.dev/v1",
            "kind": "Service",
            "metadata": {
              "annotations": {
                "client.knative.dev/user-image": "us-docker.pkg.dev/cloudrun/container/hello",
                "run.googleapis.com/client-name": "cloud-console",
                "run.googleapis.com/ingress": "all",
                "run.googleapis.com/ingress-status": "all",
                "run.googleapis.com/launch-stage": "BETA",
                "serving.knative.dev/creator": "user@email.com",
                "serving.knative.dev/lastModifier": "user@email.com",
              },
              "creationTimestamp": "2021-03-09T20:41:29.622532Z",
              "generation": 1,
              "labels": {
                "cloud.googleapis.com/location": "us-central1",
              },
              "name": "example-cloud-run-service",
              "namespace": "711888229551",
              "resourceVersion": "AAW9IJgvqVc",
              "selfLink": "/apis/serving.knative.dev/v1/namespaces/711888229551/services/example-cloud-run-service",
              "uid": "fce2d165-9ccb-4b6c-9204-67362d41aadd",
            },
            "spec": {
              "template": {
                "metadata": {
                  "annotations": {
                    "autoscaling.knative.dev/maxScale": "100",
                    "run.googleapis.com/client-name": "cloud-console",
                    "run.googleapis.com/sandbox": "gvisor",
                  },
                  "name": "example-cloud-run-service-00001-mub",
                },
                "spec": {
                  "containerConcurrency": 80,
                  "containers": [
                    {
                      "image": "us-docker.pkg.dev/cloudrun/container/hello",
                      "ports": [
                        {
                          "containerPort": 8080,
                        },
                      ],
                      "resources": {
                        "limits": {
                          "cpu": "1000m",
                          "memory": "256Mi",
                        },
                      },
                    },
                  ],
                  "serviceAccountName": "711888229551-compute@developer.gserviceaccount.com",
                  "timeoutSeconds": 300,
                },
              },
              "traffic": [
                {
                  "latestRevision": true,
                  "percent": 100,
                },
              ],
            },
            "status": {
              "address": {
                "url": "https://example-cloud-run-service-rhbpqun2qa-uc.a.run.app",
              },
              "conditions": [
                {
                  "lastTransitionTime": "2021-03-09T20:42:53.416791Z",
                  "status": "True",
                  "type": "Ready",
                },
                {
                  "lastTransitionTime": "2021-03-09T20:42:53.211458Z",
                  "status": "True",
                  "type": "ConfigurationsReady",
                },
                {
                  "lastTransitionTime": "2021-03-09T20:42:53.416791Z",
                  "status": "True",
                  "type": "RoutesReady",
                },
              ],
              "latestCreatedRevisionName": "example-cloud-run-service-00001-mub",
              "latestReadyRevisionName": "example-cloud-run-service-00001-mub",
              "observedGeneration": 1,
              "traffic": [
                {
                  "latestRevision": true,
                  "percent": 100,
                  "revisionName": "example-cloud-run-service-00001-mub",
                },
              ],
              "url": "https://example-cloud-run-service-rhbpqun2qa-uc.a.run.app",
            },
          },
        },
      ],
      "_type": "google_cloud_run_service",
      "active": undefined,
      "category": [
        "infrastructure",
      ],
      "createdOn": 1615322489622,
      "displayName": "example-cloud-run-service",
      "function": [
        "workload-management",
      ],
      "name": "example-cloud-run-service",
      "webLink": "https://console.cloud.google.com/run/detail/us-central1/example-cloud-run-service",
    },
  ],
  "collectedRelationships": [],
  "encounteredTypes": [
    "google_cloud_run_service",
  ],
  "numCollectedEntities": 1,
  "numCollectedRelationships": 0,
}
`;
